orcas/l1l2cassandra.go

```diff
@@ -0,0 +1,411 @@
+	// Set to L2
+	metrics.IncCounter(orcas.MetricCmdSetL2)
+	start := timer.Now()
+	errL2 := l.l2.Set(req)
+
+	metrics.ObserveHist(orcas.HistSetL2, timer.Since(start))
+
+	// Return an error if data is not stored in L2
+	if errL2 != nil {
+		metrics.IncCounter(orcas.MetricCmdSetErrorsL2)
+		return ErrL2OnlySetFailed
+	}
+
+	metrics.IncCounter(orcas.MetricCmdSetSuccessL2)
+	// return code is ignored by goroutine caller (Set func)
+	return errL2
+}
+
+func (l *L1L2CassandraOrca) Set(req common.SetRequest) error {
+	//log.Println("set", string(req.Key))
+
+	// Set to L1 first
+	metrics.IncCounter(orcas.MetricCmdSetL1)
+	start := timer.Now()
+	errL1 := l.l1.Set(req)
+
+	metrics.ObserveHist(orcas.HistSetL1, timer.Since(start))
+
+	if errL1 != nil {
+		metrics.IncCounter(orcas.MetricCmdSetErrorsL1)
+	} else {
+		// Successful write to L1 is a completed write !
+		metrics.IncCounter(orcas.MetricCmdSetSuccessL1)
+		metrics.IncCounter(orcas.MetricCmdSetSuccess)
+		// Set L2 asynchronously. As L1 is OK, we don't check L2 success.
+		go l.SetL2Only(req)
+		return l.res.Set(req.Opaque, req.Quiet)
+	}
+
+	// If L1 Set failed, fallback to L2 Set
+	metrics.IncCounter(orcas.MetricCmdSetL2)
+	start = timer.Now()
+	errL2 := l.l2.Set(req)
+
+	metrics.ObserveHist(orcas.HistSetL2, timer.Since(start))
+
+	// Return an error if data is not stored in L2 either
+	if errL2 != nil {
+		metrics.IncCounter(orcas.MetricCmdSetErrorsL2)
+		metrics.IncCounter(orcas.MetricCmdSetErrors)
+		return ErrL1L2SetFailed
+	}
+
+	metrics.IncCounter(orcas.MetricCmdSetSuccessL2)
+	metrics.IncCounter(orcas.MetricCmdSetSuccess)
+	return l.res.Set(req.Opaque, req.Quiet)
+}
+
+func (l *L1L2CassandraOrca) Add(req common.SetRequest) error {
+	// Add is not yet implemented.
+	log.Println("[WARN] Add command not supported by L1L2 NotConsistent orchestrator")
+	return common.ErrUnknownCmd
+}
+
+func (l *L1L2CassandraOrca) Replace(req common.SetRequest) error {
+	// Replace is not yet implemented.
+	log.Println("[WARN] Replace command not supported by L1L2 NotConsistent orchestrator")
+	return common.ErrUnknownCmd
+}
+
+func (l *L1L2CassandraOrca) Append(req common.SetRequest) error {
+	// Append is not yet implemented.
+	log.Println("[WARN] Append command not supported by L1L2 NotConsistent orchestrator")
+	return common.ErrUnknownCmd
+}
+
+func (l *L1L2CassandraOrca) Prepend(req common.SetRequest) error {
+	// Prepend is not yet implemented.
+	log.Println("[WARN] Prepend command not supported by L1L2 NotConsistent orchestrator")
+	return common.ErrUnknownCmd
+}
+
+func (l *L1L2CassandraOrca) Delete(req common.DeleteRequest) error {
+	//log.Println("delete", string(req.Key))
+
+	// Try L2 first
+	metrics.IncCounter(MetricCmdDeleteL2)
+	start := timer.Now()
+
+	err := l.l2.Delete(req)
+
+	metrics.ObserveHist(HistDeleteL2, timer.Since(start))
+
+	if err != nil {
+		// On a delete miss in L2 don't bother deleting in L1. There might be no
```

Comment:
I think we should do it in L1 too, to help the consistency if something has been going wrong.
Paying for the L1 latency should be cheap as we already has waited for the L2